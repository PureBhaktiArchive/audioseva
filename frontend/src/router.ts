/*
 * sri sri guru gauranga jayatah
 */

import firebase from "firebase/app";

import Vue from "vue";
import Router from "vue-router";

Vue.use(Router);

export const router = new Router({
  mode: "history",
  routes: [
    { path: "*", redirect: "/" },
    {
      path: "/login",
      meta: { guestOnly: true },
      component: () => import("@/views/Login.vue")
    },
    {
      path: "/",
      meta: { requireAuth: true },
      component: () => import("@/views/Dashboard.vue"),
      children: [
        {
          path: "",
          component: () => import("@/views/Home.vue")
        },
        {
          path: "cr/allot",
          component: () => import("@/views/CRAllotment.vue")
        },
        {
          path: "sqr/",
          component: () => import("@/views/SQR.vue"),
          meta: { activator: true, activatorName: "Sound Quality Reporting", menuIcon: "fas fa-headphones" },
          children: [
            { path: "",
              component: () => import("@/views/SQRFiles.vue"),
              meta: { menuItem: true}
            },
            {
              path: "allot",
              component: () => import("@/views/SQRAllotment.vue"),
              meta: { menuItem: true }
            },
            { path: "statistics",
              component: () => import("@/views/SQRFileStatistics.vue"),
              meta: { menuItem: true }
            }
          ]
        },
        {
          path: "se/",
          component: () => import("@/views/SE/SE.vue"),
          meta: { activator: true, activatorName: "Sound Engineering", menuIcon: "fas fa-music" },
          children: [
            { path: "", component: () => import("@/views/SE/Tasks.vue"), meta: { menuItem: true } },
            { path: "allot", component: () => import("@/views/SE/Allotment.vue"), meta: { menuItem: true } }
          ]
        },
        {
          path: "te/allot",
          component: () => import("@/views/TEAllotment.vue")
        },
        {
          path: "te/fc/allot",
          component: () => import("@/views/TFCAllotment.vue")
        },
        {
          path: "/qc/allot",
          component: () => import("@/views/QCAllotment.vue")
        }
      ]
    }
  ]
});

router.beforeEach((to, from, next) => {
  let currentUser = firebase.auth().currentUser;
  let requireAuth = to.matched.some(record => record.meta.requireAuth);
  let guestOnly = to.matched.some(record => record.meta.guestOnly);

  if (requireAuth && !currentUser)
    next({
      path: "/login",
      query: { redirect: to.fullPath }
    });
  else if (guestOnly && currentUser) next("/");
  else next();
});
