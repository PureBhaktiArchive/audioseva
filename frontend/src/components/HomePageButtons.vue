<template>
  <div>
    <v-btn class="mx-2" :key="`${button.text}-${index}`" v-for="(button, index) in routeButtons" v-bind="button.props">
      {{ button.text }}
    </v-btn>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import { RouteConfig } from "vue-router";
import { getHomePageRoutes } from "@/router";

@Component({
  name: "HomePageButtons"
})
export default class HomePageButtons extends Vue {
  userClaims: any = {};
  routeButtons: any[] = [];

  async mounted() {
    this.makeButtons(await getHomePageRoutes());
  }

  makeButtons(routes: RouteConfig[], parentPath: string = "/") {
    const defaultButtonProps = {
      color: "primary"
    };
    routes.forEach(route => {
      const fullPath = `${parentPath}${route.path}`;
      if (route.meta.homePageLink) {
        this.routeButtons.push({
          text: route.meta.homePageLink.text,
          props: {
            ...defaultButtonProps,
            ...route.meta.homePageLink.props,
            to: fullPath
          }
        });
      }
      if (route.children) {
        this.makeButtons(route.children, fullPath);
      }
    });
  }
}
</script>

<style scoped>
</style>
