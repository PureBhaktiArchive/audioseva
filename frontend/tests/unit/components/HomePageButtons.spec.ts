import Vue from "vue";
import { createLocalVue, shallowMount } from "@vue/test-utils";
import VueRouter from "vue-router";
import HomePageButtons from "@/components/HomePageButtons.vue";
import { router } from "@/router";

describe("HomePageButtons", () => {
  let localVue: typeof Vue;

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(VueRouter);
  });

  it("should render buttons that match TE claims", () => {
    const wrapper = shallowMount(HomePageButtons, {
      localVue,
      router,
      methods: {
        getUserClaims: () => ({ TE: true })
      }
    });
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.routeButtons).toMatchSnapshot();
    });
  });

  it("should render buttons that match coordinator claims", () => {
    const wrapper = shallowMount(HomePageButtons, {
      localVue,
      router,
      methods: {
        getUserClaims: () => ({ coordinator: true })
      }
    });
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.routeButtons).toMatchSnapshot();
    });
  });
});
