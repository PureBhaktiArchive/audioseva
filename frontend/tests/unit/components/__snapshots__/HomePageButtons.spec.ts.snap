// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageButtons should render buttons that match TE claims 1`] = `
Array [
  Object {
    "props": Object {
      "color": "primary",
      "to": "/te/upload",
    },
    "text": "Upload",
  },
]
`;

exports[`HomePageButtons should render buttons that match coordinator claims 1`] = `
Array [
  Object {
    "props": Object {
      "color": "primary",
      "to": "/sqr/",
    },
    "text": "SQR",
  },
  Object {
    "props": Object {
      "color": "primary",
      "to": "/te/tasks",
    },
    "text": "TE",
  },
]
`;
